go.sum

```diff
@@ -1,5 +1,7 @@
 github.com/davecgh/go-spew v1.1.1 h1:vj9j/u1bqnvCEfJOwUhtlOARqs3+rkHYY13jYWTU97c=
 github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
+github.com/orcaman/concurrent-map/v2 v2.0.1 h1:jOJ5Pg2w1oeB6PeDurIYf6k9PQ+aTITr/6lP/L/zp6c=
+github.com/orcaman/concurrent-map/v2 v2.0.1/go.mod h1:9Eq3TG2oBe5FirmYWQfYO5iH1q0Jv47PLaNK++uCdOM=
```

Comment:
How do you think if we use: https://github.com/allegro/bigcache
To me, it's more stable

Reply:
Ive never used [bigcache](https://github.com/allegro/bigcache) before. Reading the code, I see it implements some functionality that I manually implemented, like the expire time. Let me look into it, make sure it has all we need and update the PR with it. 

It does look more actively maintained.

Reply:
I don't think bigcache is better for this use case,  here are some reasons why :
 
1. bigcache needs a config, we can use bigcache.defaultConfig...but what if the library changes the default config between versions.
2. bigcaches uses errors all over the place, we will need to handle errors in functions where there were no error handling before. For example, calling bigcache.New(...) returns a error.  While creating a simple concurrent-map does not.
3. bigcache only supports byte slices as values. This means we need to convert pdu.PDU into a byte array. We already do when we Marshal the PDU in WrtiePdu from connection.go. So we either Marshall before we call WritePdu, or we move the window variable into connection.go. Either case, we will also need to modify Read function in connection.go.
4. When we receive the response, we will need to unmarshal  the request PDU from []byte back to pdu.PDU. A waste of resources if you ask me.

We don't really need a "big cache". I am already limiting the window size to max 255 by setting the type to uint8. If your window fills to 255, you got a serious issue on the SMSC or the ESME is abusing the connection. In either case, 255 max is a good value and a very small size for concurrent-map ans should not impact the library performance.

Reply:
@laduchesneau
The motivation for using external caching:
- Stability
- Less maintenance cost

To avoid specialized configuration to use bigcache, I highly recommend we follow Dependency Injection:
- Let user inject `cache`
- We provide default caching with bigcache default config

```go
type SubmitSMTracker interface {
   Put(context.Context, messageID int32) error
   Get(context.Context, messageID int32) Status
   List(context.Context) []Status
}
```

```go
NewTransceiver(.... , submitSMTracker SubmitSMTracker) *Transceiver {
   return &Transceiver {
         ....
         submitSMTracker: submitSMTracker
   }
}
```

```go
type DefaultSubmitSMTracker struct {
   cache *bigcache.Cache
}

func NewDefaultSubmitSMTracker(... some specialized configs) *DefaultSubmitSMTracker {
   ....
}

// implement SubmitSMTracker interface
....
```

In this way, user are free to implement their own map/caching mechanism. Some of them want to use Redis, the others might wanna use RocksDB.

Reply:
And I don't worry about marshal/unmarshal performance. I bet it's very fast, much faster than json encoding.

Reply:
I really like your idea about user implementation and letting users choose their preferred DB/Cache/Map solution. I would take it a step further and not implement a default cache. Just provide a interface.

I would also leave to the user to handle all errors of their solution. 

```
type RequestWindow interface {
	Set(context.Context, Request)
	Get(context.Context, int32) (Request, bool)
	List(context.Context) []Request
	Delete(context.Context, int32)
	Clear(context.Context)
	Length(context.Context) int32
}
```

Reply:
@laduchesneau I just tried as per your suggestion for using interface,
https://github.com/laduchesneau/gosmpp/compare/master...sujit-baniya:gosmpp-forked:master